Dalamud/Interface/Textures/Internal/TextureManager.Wic.cs

```diff
@@ -0,0 +1,656 @@
+using Dalamud.Plugin.Services;
+using Dalamud.Utility;
+using Dalamud.Utility.TerraFxCom;
+
+using TerraFX.Interop.DirectX;
+using TerraFX.Interop.Windows;
+
+using static TerraFX.Interop.Windows.Windows;
+
+namespace Dalamud.Interface.Textures.Internal;
+
+/// <summary>Service responsible for loading and disposing ImGui texture wraps.</summary>
+[SuppressMessage(
+    "StyleCop.CSharp.LayoutRules",
+    "SA1519:Braces should not be omitted from multi-line child statement",
+    Justification = "Multiple fixed blocks")]
+internal sealed partial class TextureManager
+{
+    /// <inheritdoc/>
+    public async Task SaveToStreamAsync(
+        IDalamudTextureWrap? wrap,
+        Guid containerGuid,
+        Stream? stream,
+        IReadOnlyDictionary<string, object>? props = null,
+        bool leaveWrapOpen = false,
+        bool leaveStreamOpen = false,
+        CancellationToken cancellationToken = default)
+    {
+        try
+        {
+            if (wrap is null)
+                throw new NullReferenceException($"{nameof(wrap)} cannot be null.");
+            if (stream is null)
+                throw new NullReferenceException($"{nameof(stream)} cannot be null.");
+
+            using var istream = ManagedIStream.Create(stream, true);
+            using var wrapAux = new WrapAux(wrap, true);
+
+            var dxgiFormat =
+                WicManager.GetCorrespondingWicPixelFormat(wrapAux.Desc.Format, out _, out _)
+                    ? wrapAux.Desc.Format
+                    : DXGI_FORMAT.DXGI_FORMAT_B8G8R8A8_UNORM;
+
+            var (specs, bytes) = await this.GetRawImageAsync(wrapAux, new() { Format = dxgiFormat }, cancellationToken)
+                                           .ConfigureAwait(false);
+
+            await Task.Run(
+                () => this.Wic.SaveToStreamUsingWic(
+                    specs,
+                    bytes,
+                    containerGuid,
+                    istream,
+                    props,
+                    cancellationToken),
+                cancellationToken);
+        }
+        finally
+        {
+            if (!leaveWrapOpen)
+                wrap?.Dispose();
+            if (!leaveStreamOpen)
+                stream?.Dispose();
+        }
+    }
+
+    /// <inheritdoc/>
+    public async Task SaveToFileAsync(
+        IDalamudTextureWrap? wrap,
+        Guid containerGuid,
+        string? path,
+        IReadOnlyDictionary<string, object>? props = null,
+        bool leaveWrapOpen = false,
+        CancellationToken cancellationToken = default)
+    {
+        try
+        {
+            if (wrap is null)
+                throw new NullReferenceException($"{nameof(wrap)} cannot be null.");
+            if (path is null)
+                throw new NullReferenceException($"{nameof(path)} cannot be null.");
+
+            using var wrapAux = new WrapAux(wrap, true);
+            var pathTemp = $"{path}.{GetCurrentThreadId():X08}{Environment.TickCount64:X16}.tmp";
```

Comment:
Is this safe? In my experience this kinda code always blows up. Maybe we need something like this? [GetTempPath](https://github.com/goatcorp/FFXIVQuickLauncher/blob/6e0829a973b26db3aa95a9e60b96d73ff222a1ae/src/XIVLauncher.Common/Util/PlatformHelpers.cs#L34)

Reply:
Creating a file in a temporary directory may end up creating a file in a different drive, so that move/replace operation cannot happen without copying. While I'm expecting this to be safe (thread-unique, effectively monotonic counter) to be safe and to offer a single point that needs to be fixed in case of errors, I'll add a utility function for making a temporary file name that should be able to be used for file replace operations.